tensorboard/data/provider.py

```diff
@@ -33,6 +33,25 @@ class DataProvider(object):
   downsampling strategies or domain restriction by step or wall time.
   """
 
+  @abc.abstractmethod
+  def list_runs(self, experiment_id):
+    """List all runs within an experiment.
+
+    The returned sequence should be sorted by run start time (i.e., the
+    time of the first event in each run), so that as new runs are
+    created under a running TensorBoard the list of runs is append-only,
+    and thus each run's time series color stays constant over time. In
+    case two runs have the same start time, their relative order should
+    be tie-broken deterministically (e.g., by run name).
+
+    Args:
+      experiment_id: ID of enclosing experiment.
+
+    Returns:
+      A sequence of distinct run names, sorted as described above.
```

Comment:
No AI required: of course, not related to your code but I definitely want FE and BE to communicate via ids, not names in the future. WDYT? For local FS based implementation, the name can be the id 🤷‍♂ .

Reply:
Good question; I also wondered that while writing these docs.

Definitely agree that communication should primarily be ID-based.

This API call probably should return the run names, because the frontend
needs to display them, but it could return `(id, name)` structs. Perhaps
the real place to ask this question is on `RunTagFilter`’s definition,
which is currently specified in terms of run and tag names, and should
probably be changed to run and tag IDs.

To properly make this change, we’d need to plumb the IDs through to the
frontend, so that requests to `/data/plugin/scalars/tags` can specify
the run IDs as query parameters. Just for the purpose of unblocking
internal work, I think that I’d prefer to keep these as run names for
now, though I could be convinced otherwise. These APIs are clearly and
explicitly marked experimental, so I will happily make breaking changes
to them when the time comes.